training/src/main/scala/com/airbnb/aerosolve/training/Evaluation.scala

```diff
@@ -20,119 +20,25 @@ object Evaluation {
-      val tmpMap = tmp.toMap
-      val f1 = tmpMap.getOrElse(evalMetric, 0.0)
-      if (f1 > bestF1) {
-        bestF1 = f1
-        metrics = tmp
-      }
-      trainPR.append((tmpMap.getOrElse("!TRAIN_PRECISION", 0.0), tmpMap.getOrElse("!TRAIN_RECALL", 0.0), threshold))
-      holdPR.append((tmpMap.getOrElse("!HOLD_PRECISION", 0.0), tmpMap.getOrElse("!HOLD_RECALL", 0.0), threshold))
-    }
-
-    metricsAppend(metrics, trainPR, holdPR)
-
-    evaluateBinaryClassificationAUC(records, metrics)
-
-    metrics
-      .sortWith((a, b) => a._1 < b._1)
-      .toArray
+    evaluateBinaryClassificationWithResults(records, buckets, evalMetric, null)
   }
 
-  def evaluateBinaryClassification(records : RDD[EvaluationRecord],
+  def evaluateBinaryClassificationWithResults(records : RDD[EvaluationRecord],
                                    buckets : Int,
                                    evalMetric : String,
                                    resultsOutputPath: String) : Array[(String, Double)] = {
-    var metrics  = mutable.Buffer[(String, Double)]()
-    var bestF1 = -1.0
-    val thresholds = records.map(x => x.score).histogram(buckets)._1
-    var results = mutable.Buffer[(String, Double)]()
-    var holdThresholdPrecisionRecall = mutable.Buffer[(Double, Double, Double)]()
-    var trainThresholdPrecisionRecall = mutable.Buffer[(Double, Double, Double)]()
-
-    // Search all thresholds for the best F1
-    // At the same time collect the precision and recall.
-    val trainPR = new ArrayBuffer[(Double, Double, Double)]()
-    val holdPR = new ArrayBuffer[(Double, Double, Double)]()
-    trainPR.append((1.0, 0.0, thresholds.max))
-    holdPR.append((1.0, 0.0, thresholds.max))
-
-    for (i <- 0 until thresholds.size - 1) {
-      val threshold = thresholds(i)
-      val tmp = evaluateBinaryClassificationAtThreshold(records, threshold, results)
-      val tmpMap = tmp.toMap
-      val f1 = tmpMap.getOrElse(evalMetric, 0.0)
-      if (f1 > bestF1) {
-        bestF1 = f1
-        metrics = tmp
-      }
-      trainPR.append((tmpMap.getOrElse("!TRAIN_PRECISION", 0.0), tmpMap.getOrElse("!TRAIN_RECALL", 0.0), threshold))
-      holdPR.append((tmpMap.getOrElse("!HOLD_PRECISION", 0.0), tmpMap.getOrElse("!HOLD_RECALL", 0.0), threshold))
-    }
-
-    metricsAppend(metrics, trainPR, holdPR, results)
-    evaluateBinaryClassificationAUC(records, metrics, results)
-
-    // Write results
-    thresholdsAppend(trainPR, holdPR, holdThresholdPrecisionRecall, trainThresholdPrecisionRecall)
-    val resultsArray = results.sortWith((a, b) => a._1 < b._1).toArray
-    ResultUtil.writeResults(resultsOutputPath, resultsArray, holdThresholdPrecisionRecall.toArray, trainThresholdPrecisionRecall.toArray)
-
-    metrics
-      .sortWith((a, b) => a._1 < b._1)
-      .toArray
+    // Convert RDD to list
+    val recordsList = List.fromArray(records.collect());
+    evaluateBinaryClassificationWithResults(recordsList, buckets, evalMetric, resultsOutputPath)
```

Comment:
I see. It makes sense but I'm having trouble collecting them to driver and driver is struggling with the evaluation. I'm going to try making everything follow the RDD codepath instead.